torch/_dynamo/guards.py

```diff
@@ -501,13 +504,20 @@ def TENSOR_MATCH(self, guard: Guard):
             # as an empty set is a safe degeneration - that is, a strictly static tensor is always valid for a frame
             # compiled with that same
             # tensor + more onerous user directives.
+            #
+            # There is finer grain control available within _dynamo_dynamic_indices - and that is for ranges.
+            # To fully understand that, please see the mark_dynamic_constrain api docs first.
+            # The way we guard on user directive ranges specified via mark_dynamic_constrain is by the same logic as
+            # above. Specifically, when X is a strict subset of Y notion around which dimensions can be valid, the same
+            # goes for the value ranges WITHIN a dimension. We specialize to the user directive range. Ranges within
+            # that range can safely reuse the same compiled frame, but wider ranges must recompile.
             static, reason = tensor_always_has_static_shape(
                 value, guard.source, is_tensor=True
             )
             if not static:
                 if hasattr(value, "_dynamo_dynamic_indices"):
                     code.append(
-                        f"({tensor_name}._dynamo_dynamic_indices.issubset({value._dynamo_dynamic_indices})) if hasattr({tensor_name}, '_dynamo_dynamic_indices') else True"  # noqa: B950
+                        f"___dynamic_dims_check({tensor_name}, {dict(value._dynamo_dynamic_indices)})"
```

Comment:
This is starting to look expensive. What if we do some sort of counter check (and when you update constraints, we bump the counter)?

Reply:
Sorry, I don't understand this.

Reply:
Yes. Update when we change metadata

Reply:
Well, then you will never reuse past frames, right? That would be fine for invalidating the last frame, but it would be cache inefficient. 

I think we can move this to C++ though, since its per tensor, do you want me to follow up with that?